[TIP] Run Playwright Tests on unsupported Linux distributions

To run Playwright on unsupported Linux distribution, you run run Playwright server in Docker while keeping your tests running on the host system:

### Run Playwright Server in Docker
```bash
docker run -p 3000:3000 --rm --init -it mcr.microsoft.com/playwright:v1.41.0-jammy /bin/sh -c "cd /home/pwuser && npx -y playwright@1.41.0 run-server --port 3000 --host 0.0.0.0"
```

It will output following when ready
```
Listening on ws://127.0.0.1:3000/
```

###  Point Playwright client to the server on Docker

If running @playwright/test, use environment variable, no other changes are needed.

```bash
PW_TEST_CONNECT_WS_ENDPOINT=ws://127.0.0.1:3000/ npx playwright test
```

Alternatively, for other applications and language ports, use [browserType.connect](https://playwright.dev/docs/api/class-browsertype#browser-type-connect) API to use remote Playwright.

```js
const browser = await playwright['chromium'].connect('ws://127.0.0.1:3000/');
```

### Network

Make sure that the host network is available in your Docker container if you are testing local servers:

```sh
docker run --add-host=hostmachine:host-gateway
```

This will make `hostmachine` point to the host's localhost. Note that all the tests would also need to use `hostmachine` in the URLs to access servers on the host. Alternatively, you can follow unconventional route and do `docker run --add-host=localhost:host-gateway` at your own risk.

I tried to run Java tests with the code as in the https://github.com/microsoft/playwright/issues/26482#issuecomment-1869288427 but I'm getting an error running `gradle test`:

```
LoginTest > initializationError FAILED
    com.microsoft.playwright.PlaywrightException at LoginTest.java:37
        Caused by: com.microsoft.playwright.impl.DriverException at LoginTest.java:37
```
The row 37 is` Browser browser = playwright.chromium().connect("ws://127.0.0.1:3000/");`

Also I'm not sure if I'm using a correct image as here:
`docker run -p 3000:3000 --rm --init -it mcr.microsoft.com/playwright:v1.41.0-jammy /bin/sh -c "cd /home/pwuser && npx -y playwright@1.41.0 run-server --port 3000 --host 0.0.0.0"`
vs 
`mcr.microsoft.com/playwright/**java**:v1.41.0-jammy`
 But the 
`docker run -p 3000:3000 --rm --init -it mcr.microsoft.com/playwright/java:v1.41.0-jammy /bin/sh -c "cd /home/pwuser && npx -y playwright@1.41.0 run-server --port 3000 --host 0.0.0.0"`
complains on `/bin/sh: 1: npx: not found`

Any help is much appreciated.

I tried to run Java tests with the code as in the https://github.com/microsoft/playwright/issues/26482#issuecomment-1869288427 but I'm getting an error running `gradle test`:

```
LoginTest > initializationError FAILED
    com.microsoft.playwright.PlaywrightException at LoginTest.java:37
        Caused by: com.microsoft.playwright.impl.DriverException at LoginTest.java:37
```
The row 37 is` Browser browser = playwright.chromium().connect("ws://127.0.0.1:3000/");`

Also I'm not sure if I'm using a correct image as here:
`docker run -p 3000:3000 --rm --init -it mcr.microsoft.com/playwright:v1.41.0-jammy /bin/sh -c "cd /home/pwuser && npx -y playwright@1.41.0 run-server --port 3000 --host 0.0.0.0"`
vs 
`mcr.microsoft.com/playwright/**java**:v1.41.0-jammy`
 But the 
`docker run -p 3000:3000 --rm --init -it mcr.microsoft.com/playwright/java:v1.41.0-jammy /bin/sh -c "cd /home/pwuser && npx -y playwright@1.41.0 run-server --port 3000 --host 0.0.0.0"`
complains on `/bin/sh: 1: npx: not found`

Any help is much appreciated.

I tried to run Java tests with the code as in the https://github.com/microsoft/playwright/issues/26482#issuecomment-1869288427 but I'm getting an error running `gradle test`:

```
LoginTest > initializationError FAILED
    com.microsoft.playwright.PlaywrightException at LoginTest.java:37
        Caused by: com.microsoft.playwright.impl.DriverException at LoginTest.java:37
```
The row 37 is` Browser browser = playwright.chromium().connect("ws://127.0.0.1:3000/");`

Also I'm not sure if I'm using a correct image as here:
`docker run -p 3000:3000 --rm --init -it mcr.microsoft.com/playwright:v1.41.0-jammy /bin/sh -c "cd /home/pwuser && npx -y playwright@1.41.0 run-server --port 3000 --host 0.0.0.0"`
vs 
`mcr.microsoft.com/playwright/**java**:v1.41.0-jammy`
 But the 
`docker run -p 3000:3000 --rm --init -it mcr.microsoft.com/playwright/java:v1.41.0-jammy /bin/sh -c "cd /home/pwuser && npx -y playwright@1.41.0 run-server --port 3000 --host 0.0.0.0"`
complains on `/bin/sh: 1: npx: not found`

Any help is much appreciated.

--- a/docs/src/docker.md
+++ b/docs/src/docker.md
@@ -103,6 +103,88 @@ Using `--ipc=host` is recommended when using Chrome ([Docker docs](https://docs.
 
 See our [Continuous Integration guides](./ci.md) for sample configs.
 
+### Remote Connection
+
+You can run Playwright Server in Docker while keeping your tests running on the host system or another machine. This is useful for running tests on unsupported Linux distributions or remote execution scenarios.
+
+#### Running the Playwright Server
+
+Start the Playwright Server in Docker:
+
+```bash
+docker run -p 3000:3000 --rm --init -it --workdir /home/pwuser --user pwuser mcr.microsoft.com/playwright:v%%VERSION%%-noble /bin/sh -c "npx -y playwright@%%VERSION%% run-server --port 3000 --host 0.0.0.0"
+```
+
+#### Connecting to the Server
+* langs: js
+
+There are two ways to connect to the remote Playwright server:
+
+1. Using environment variable with `@playwright/test`:
+
+```bash
+PW_TEST_CONNECT_WS_ENDPOINT=ws://127.0.0.1:3000/ npx playwright test
+```
+
+2. Using the [`method: BrowserType.connect`] API for other applications:
+
+```js
+const browser = await playwright['chromium'].connect('ws://127.0.0.1:3000/');
+```
+
+#### Connecting to the Server
+* langs: python, csharp, java
+
+```python sync
+from playwright.sync_api import sync_playwright
+
+with sync_playwright() as p:
+    browser = p.chromium.connect("ws://127.0.0.1:3000/")
+```
+
+```python async
+from playwright.async_api import async_playwright
+
+async with async_playwright() as p:
+    browser = await p.chromium.connect("ws://127.0.0.1:3000/")
+```
+
+```csharp
+using Microsoft.Playwright;
+
+using var playwright = await Playwright.CreateAsync();
+await using var browser = await playwright.Chromium.ConnectAsync("ws://127.0.0.1:3000/");
+```
+
+```java
+package org.example;
+
+import com.microsoft.playwright.*;
+import java.nio.file.Paths;
+
+public class App {
+  public static void main(String[] args) {
+    try (Playwright playwright = Playwright.create()) {
+      Browser browser = playwright.chromium().connect("ws://127.0.0.1:3000/");
+    }
+  }
+}
+```
+
+#### Network Configuration
+
+If you need to access local servers from within the Docker container:
+
+```bash
+docker run --add-host=hostmachine:host-gateway -p 3000:3000 --rm --init -it --workdir /home/pwuser --user pwuser mcr.microsoft.com/playwright:v%%VERSION%%-noble /bin/sh -c "npx -y playwright@%%VERSION%% run-server --port 3000 --host 0.0.0.0"
+```
+
+This makes `hostmachine` point to the host's localhost. Your tests should use `hostmachine` instead of `localhost` when accessing local servers.
+
+:::note
+When running tests remotely, ensure the Playwright version in your tests matches the version running in the Docker container.
+:::
+
 ## Image tags
 
 See [all available image tags].

--- a/packages/playwright-core/src/cli/program.ts
+++ b/packages/playwright-core/src/cli/program.ts
@@ -291,7 +291,7 @@ program
     });
 
 program
-    .command('run-server', { hidden: true })
+    .command('run-server')
     .option('--port <port>', 'Server port')
     .option('--host <host>', 'Server host')
     .option('--path <path>', 'Endpoint Path', '/')

